File: go/array_hashing/242_valid_anagram.go
package main

import (
	"fmt"
)

func main() {
	// test cases

	// result: true
	fmt.Println(isAnagram("anagram", "nagaram"))
	// result: false
	fmt.Println(isAnagram("rat", "car"))
}

func isAnagram(s string, t string) bool {
	// hashmap approach
	if len(s) != len(t) {
		return false
	}

	charCount := make(map[rune]int)

	// convert string to runes array
	sChars := []rune(s)
	// save runes in map with char count
	for i := 0; i < len(sChars); i++ {
		if _, exists := charCount[sChars[i]]; exists {
			charCount[sChars[i]] += 1
			continue
		}
		charCount[sChars[i]] = 1
	}

	tChars := []rune(t)
	for i := 0; i < len(tChars); i++ {
		if _, exists := charCount[tChars[i]]; exists {
			charCount[tChars[i]] -= 1
		} else {
			// if character doesn't exists in original string, it's not valid
			return false
		}
	}

	for _, v := range charCount {
		if v > 0 {
			// return false if there are still some chars with value greater than 0
			return false
		}
	}

	return true
}

File: js/array_hashing/242_valid_anagram.ts
function isAnagram(s: string, t: string): boolean {
    if (s.length > t.length || t.length > s.length) return false

    const charsCount: Map<string, number> = new Map()

    for (const char of s) {
        charsCount.get(char) ? charsCount.set(char, (charsCount.get(char) || 0) + 1) : charsCount.set(char, 1)
    }

    for (const char of t) {
        if (charsCount.get(char)) {
            charsCount.set(char, (charsCount.get(char) || 0) - 1)
            continue
        }
        return false
    }

    for (const [_, count] of charsCount) {
        if (count > 0) return false
    }

    return true;
}

const testCases = () => {
    // Test cases

    // result: true
    // performance.mark('startTestCase1')
    console.log(isAnagram('anagram', 'nagaram'))
    // performance.mark('endTestCase1')
    // console.log(performance.measure('startTestCase1', 'endTestCase1'))

    // result: true
    console.log(isAnagram('rat', 'car'))
};

testCases()
